pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/metadata/ColumnMetadataImpl.java

```diff
@@ -67,14 +70,19 @@ public class ColumnMetadataImpl implements ColumnMetadata {
   private final int _totalNumberOfEntries;
   private final PartitionFunction _partitionFunction;
   private final Set<Integer> _partitions;
-  private final Map<IndexType<?, ?, ?>, Long> _indexSizeMap;
+
+  /* List of longs, each encodes:
+   * 2 byte - numeric id of IndexType
+   * 6 byte - index size */
+  private final LongArrayList _indexTypeSizeList;
```

Comment:
Same as in pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java. This is a change I find interesting, but it makes the code more fragile.

I recommend creating a new class that encapsulates the same logic while keeping ColumnMetadataImpl and SegmentMetadataImpl clean. In this case, the new code is not so aggressive in terms of new lines and state management, so consider this an optional change.